tools/ast_gen.cpp

```diff
@@ -28,33 +30,92 @@ struct Node {
 
 int main() {
   std::vector<Node> nodes{
-      {"Binary", {{"Expr", "left"}, {"Token", "operatr"}, {"Expr", "right"}}},
-      {"Grouping", {{"Expr", "expression"}}},
-      {"Unary", {{"Token", "operatr"}, {"Expr", "right"}}}};
+      {"Binary",
+       {{"std::unique_ptr<Expr>", "lhs"},
+        {"Token", "op"},
+        {"std::unique_ptr<Expr>", "rhs"}}},
+      {"Grouping", {{"std::unique_ptr<Expr>", "expr"}}},
+      {"String", {{"std::string", "value"}}},
+      {"Number", {{"double", "value"}}},
+      {"Boolean", {{"bool", "value"}}},
+      {"Nil", {{"std::nullptr_t", "value"}}},
```

Comment:
I claim `Nil` doesn't need any members since it it technically stateless.
```suggestion
      {"Nil", {}},
```